core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java

```diff
@@ -29,120 +31,95 @@ public class UnavailableSpaceBuilder {
             BlockInfra blockInfra,
-            int route,
-            double timeStart,
-            double timeEnd,
-            long distanceStart,
-            long distanceEnd
+            Collection<ResultTrain.SpacingRequirement> requirements,
+            double marginToAddBeforeEachBlock,
+            double marginToAddAfterEachBlock
     ) {
-        var routeList = new MutableStaticIdxArrayList<Route>();
-        routeList.add(route);
-        for (var blockPath : BlockRecoveryKt.recoverBlocks(infra, blockInfra, routeList, null)) {
-            var blockList = BlockRecoveryKt.toBlockList(blockPath);
-            long routeOffset = 0;
-            for (var blockId : toIntList(blockList)) {
-                var blockLength = blockInfra.getBlockLength(blockId);
-                var start = Math.max(0, distanceStart - routeOffset);
-                var end = Math.min(blockLength, distanceEnd - routeOffset);
-                if (start < end) {
-                    var newSegment = new OccupancySegment(timeStart, timeEnd, start, end);
-                    res.put(blockId, newSegment);
-                }
-                routeOffset += blockLength;
+        var res = new HashMap<Integer, RangeSet<Double>>();
+        for (var requirement : requirements) {
+            var zoneId = infra.getZoneFromName(requirement.zone);
+            var timeRange = Range.closed(
+                    requirement.beginTime - marginToAddBeforeEachBlock,
+                    requirement.endTime + marginToAddAfterEachBlock
+            );
+            for (var blockId : toIntList(blockInfra.getBlocksInZone(zoneId))) {
+                res.putIfAbsent(blockId, TreeRangeSet.create());
+                res.get(blockId).add(timeRange);
             }
         }
+        return res;
     }
 
-    /** Returns the routes that lead into the given one */
-    private static Set<Integer> getPreviousRoutes(RawSignalingInfra infra, int route) {
-        var entry = InterlockingInfraKt.getRouteEntry(infra, route);
-        return new HashSet<>(toIntList(infra.getRoutesEndingAtDet(entry)));
+    /** Returns the blocks that lead into the given one */
+    private static Set<Integer> getPreviousBlocks(RawSignalingInfra infra, BlockInfra blockInfra, int blockId) {
+        var entry = BlockInfraImplKt.getBlockEntry(blockInfra, infra, blockId);
+        return new HashSet<>(toIntList(blockInfra.getBlocksEndingAtDetector(entry)));
     }
 
-    /** Returns the routes that follow the given one */
-    private static Set<Integer> getNextRoutes(RawSignalingInfra infra, int route) {
-        var exit = InterlockingInfraKt.getRouteExit(infra, route);
-        return new HashSet<>(toIntList(infra.getRoutesStartingAtDet(exit)));
+    /** Returns the blocks that follow the given one */
+    private static Set<Integer> getNextBlocks(RawSignalingInfra infra, BlockInfra blockInfra, int blockId) {
+        var entry = BlockInfraImplKt.getBlockExit(blockInfra, infra, blockId);
```

Comment:
this is a type confusion bug: getBlockExit is in fact getRouteExit, and was renamed for some reason, so it takes a routeId

Reply:
This is actually the correct call, it calls the `BlockInfra` version which does get the exit detector of a block.

The accidental rename of `getRouteExit` has surprisingly nothing to do with this.